src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/DateTimeMarshaler.cs

```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices.JavaScript
+{
+    public class DateTimeMarshaler
+    {
+        public static string FromJavaScriptPreFilter { get; } = "value.toISOString()";
+        public static string ToJavaScriptPostFilter { get; } = "new Date(value)";
+
+        public static DateTime FromJavaScript (string s)
+        {
+            return DateTime.Parse(s).ToUniversalTime();
+        }
+
+        public static string ToJavaScript (ref DateTime dt)
+        {
+            return dt.ToString("o");
+        }
+    }
+
+    public class DateTimeOffsetMarshaler
```

Comment:
Same problems as with DateTime class

Reply:
This is where the limitation of one string or double really hits
DateTimeOffset should probably be marshalled as two doubles both ways (on for the DateTime and one for the DateTimeOffset, both as ticks). And there is no easy way to do that without supporting multiple values/arrays.

Reply:
DateTimeOffset has no JS type to marshal to. It previously marshaled to a Date and stripped the time zone, which is what this does. If we want to break backwards compatibility, I can figure out a way to preserve the time zone but it will no longer function as a JS Date on the other end.

Reply:
Date has a timeZoneOffset in minutes that could be used in this case. But the real point i was trying to make is that if the goal is to be able to marshal value types generally this pr doesn't get there.

Reply:
Date.timeZoneOffset isn't actually the timezone offset of the date, though. It just takes the JS Date (which afaik is always UTC internally - or at least equivalent to utc internally, because js dates do not store a time zone) and does a time zone conversion to tell you the offset of your system clock from UTC:

```The number of minutes returned by getTimezoneOffset() is positive if the local time zone is behind UTC, and negative if the local time zone is ahead of UTC. For example, for UTC+10, -600 will be returned.```

Essentially, the only meaningful ways to marshal a Date are as UTC or as local time. Since they are canonically UTC, I believe marshaling them as UTC is correct: ```A JavaScript date is fundamentally specified as the number of milliseconds that have elapsed since midnight on January 1, 1970, UTC. ```

Reply:
I concur about unix-time and no timezone in JS. 
JS also doesn't have timezone info database, so it could not translate between timezones (and daylight saving or leap time).

I wonder if we should actually do the conversion via valueOf() number instead of ISO string as default. Should be faster and cheaper and always UTC. `DateTimeOffset.FromUnixTimeMilliseconds` can do that.

@lewing I would benefit from more detailed explanation of "marshal value types generally this pr doesn't get there", thank you.

Reply:
I'm not certain whether valueOf's precision is equivalent to the ISO string's, I'll have to check. If they're equivalent we can probably just use the number (though I'm not certain whether Date.valueOf is a unix time, I think it might have a slightly different epoch)